### PR TITLE
[AppVerifier] Initial skeleton

### PR DESCRIPTION
**Pull request by learn-more**
## Purpose

This PR contains an initial skeleton for application verifier.
It is still very green, and is mainly based on logging.
Most providers probably won't work, since almost everything required for them to work is still a stub.
(All exports are unimplemented, and thread + module entrypoints are not hooked, as well as our lack of debug code with CS prevents all the good stuff to actually do something).

### CODE REVIEW

File: dll/ntdll/ldr/verifier.c [hunk 1]
```diff
@@ -0,0 +1,685 @@
...
+            DbgPrint("AVRF: chain: searching in %wZ\n", &Provider->DllName);
+
+        for (DllDescriptor = Provider->ProviderDlls; DllDescriptor && DllDescriptor->DllName; ++DllDescriptor)
+        {
+            PRTL_VERIFIER_THUNK_DESCRIPTOR ThunkDescriptor;
+
+            if (AVrfpDebug & RTL_VRF_DBG_SHOWCHAINING_DEBUG)
+                DbgPrint("AVRF: chain: dll: %ws\n", DllDescriptor->DllName);
+
+            if (!_wcsicmp(DllDescriptor->DllName, DllName))
```
Comment (ThFabba) [inline dll/ntdll/ldr/verifier.c]: Could invert this and use `continue` to avoid some nesting.

File: dll/ntdll/ldr/verifier.c [hunk 2]
```diff
@@ -0,0 +1,685 @@
...
+
+    for (Entry = AVrfpVerifierProvidersList.Flink; Entry != &AVrfpVerifierProvidersList; Entry = Entry->Flink)
+    {
+        PRTL_VERIFIER_DLL_DESCRIPTOR DllDescriptor;
+
+        Provider = CONTAINING_RECORD(Entry, VERIFIER_PROVIDER, ListEntry);
+
+        for (DllDescriptor = Provider->ProviderDlls; DllDescriptor && DllDescriptor->DllName; ++DllDescriptor)
+        {
+            if (!(DllDescriptor->DllFlags & VERIFIER_DLL_FLAGS_RESOLVED) &&
```
Comment (ThFabba) [inline dll/ntdll/ldr/verifier.c]: Invert condition & use `continue` to avoid a nesting level?

File: dll/ntdll/ldr/verifier.c [hunk 3]
```diff
@@ -0,0 +1,685 @@
...
+
+                Provider = CONTAINING_RECORD(Entry, VERIFIER_PROVIDER, ListEntry);
+                for (DllDescriptor = Provider->ProviderDlls; DllDescriptor && DllDescriptor->DllName; ++DllDescriptor)
+                {
+                    PRTL_VERIFIER_THUNK_DESCRIPTOR ThunkDescriptor;
+
+                    for (ThunkDescriptor = DllDescriptor->DllThunks; ThunkDescriptor && ThunkDescriptor->ThunkName; ++ThunkDescriptor)
+                    {
+                        /* Just compare function addresses, the loader will have handled forwarders and ordinals for us */
+                        if ((PVOID)FirstThunk->u1.Function == ThunkDescriptor->ThunkOldAddress)
```
Comment (ThFabba) [inline dll/ntdll/ldr/verifier.c]: Invert condition & use `continue` to avoid a nesting level?

File: dll/ntdll/ldr/verifier.c [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+{
+    NTSTATUS Status;
+    PVERIFIER_PROVIDER Provider;
+    PLIST_ENTRY Entry;
+    WCHAR* Ptr, *Next;
+
+    Status = RtlInitializeCriticalSection(&AVrfpVerifierLock);
+    InitializeListHead(&AVrfpVerifierProvidersList);
+
+    if (!SUCCEEDED(Status))
```
Comment (ThFabba) [inline dll/ntdll/ldr/verifier.c]: NT_SUCCESS

File: dll/ntdll/ldr/verifier.c [hunk 5]
```diff
@@ -0,0 +1,685 @@
...
+        return STATUS_NO_MEMORY;
+
+    RtlInitUnicodeString(&Provider->DllName, L"verifier.dll");
+    InsertTailList(&AVrfpVerifierProvidersList, &Provider->ListEntry);
+
+    Next = AVrfpVerifierDllsString;
+
+    do
+    {
+        while (*Next == ' ' || *Next == '\t')
```
Comment (ThFabba) [inline dll/ntdll/ldr/verifier.c]: These literals could use some `L`s

File: dll/win32/verifier/verifier.c [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+    for (; ImportDescriptor->Name && ImportDescriptor->OriginalFirstThunk; ImportDescriptor++)
+    {
+        PIMAGE_THUNK_DATA OriginalThunk, FirstThunk;
+
+        OriginalThunk = (PIMAGE_THUNK_DATA)(DllBase + ImportDescriptor->OriginalFirstThunk);
+        FirstThunk = (PIMAGE_THUNK_DATA)(DllBase + ImportDescriptor->FirstThunk);
+
+        for (; OriginalThunk->u1.AddressOfData && FirstThunk->u1.Function; OriginalThunk++, FirstThunk++)
+        {
+            if (!IMAGE_SNAP_BY_ORDINAL(OriginalThunk->u1.AddressOfData))
```
Comment (ThFabba) [inline dll/win32/verifier/verifier.c]: Can avoid nesting

**Comment by HBelusca** [comment]
I'm not a big fan of putting this new verifier dll in the dll/win32 "dumping ground" ^^ Maybe inside a dll/verifier instead?

File: dll/ntdll/ldr/verifier.c (hunk 4 above)
Comment (learn-more) [reply dll/ntdll/ldr/verifier.c]: Ha shit I did it again.

Thanks for reviewing :)

**Comment by HBelusca** [comment]
Ok, if you have just one dll then no problem (I don't know how many dlls/exes/libs/... there should be at the end for the verifier implementation)

**Comment by learn-more** [comment]
It's just this core dll, the rest of the DLL's come from verifier that is in the debugging tools for windows.

**Pull request merged**